gdx/src/com/badlogic/gdx/files/FileHandle.java

```diff
@@ -416,25 +401,48 @@ public void writeBytes (byte[] bytes, int offset, int length, boolean append) {
 		}
 	}
 
-	/** Returns the paths to the children of this directory. Returns an empty list if this file handle represents a file and not a
-	 * directory. On the desktop, an {@link FileType#Internal} handle to a directory on the classpath will return a zero length
-	 * array.
-	 * @throws GdxRuntimeException if this file is an {@link FileType#Classpath} file. */
-	public FileHandle[] list () {
-		if (type == FileType.Classpath) throw new GdxRuntimeException("Cannot list a classpath directory: " + file);
-		String[] relativePaths = file().list();
-		if (relativePaths == null) return new FileHandle[0];
-		FileHandle[] handles = new FileHandle[relativePaths.length];
-		for (int i = 0, n = relativePaths.length; i < n; i++)
-			handles[i] = child(relativePaths[i]);
-		return handles;
-	}
-
-	/** Returns the paths to the children of this directory that satisfy the specified filter. Returns an empty list if this file
-	 * handle represents a file and not a directory. On the desktop, an {@link FileType#Internal} handle to a directory on the
-	 * classpath will return a zero length array.
-	 * @param filter the {@link FileFilter} to filter files
-	 * @throws GdxRuntimeException if this file is an {@link FileType#Classpath} file. */
+	 /**
+	  * Returns the paths to the children of this directory. Returns an empty list if this file handle represents a file and not a
+	  * directory. An {@link FileType#Internal} handle to a directory on the classpath will search for
+	  * an asset.index file in the * root of the asset path. This index should be generated at compile time by a gradle task.
+	  * See for reference: * https://gist.github.com/abueide/e86ea0111025233fb6493dfbcc46501e
+	  *
+	  * @throws GdxRuntimeException if this file is an {@link FileType#Classpath} file.
+	  */
+	 public FileHandle[] list () {
+		  if (type == FileType.Classpath)
+				throw new GdxRuntimeException("Cannot list a classpath directory: " + file);
+		  if (type == FileType.Internal) {
+				if (!Gdx.files.internal("assets.index").exists())
+					 //How is this supposed to be done? Is there a framework level logging system to log a warning?
+					 Gdx.app.log("WARNING: ","Asset index not found, please update your gradle configuration "
+						 + "according to: https://gist.github.com/abueide/e86ea0111025233fb6493dfbcc46501e");
+				else {
+					 String[] filePaths = Gdx.files.internal("assets.index").readString().split("\n");

```

Comment:
It would probably be more efficient to use an explicit `BufferedReader` and read the file line by line. Also I am not sure if this shouldn't be cached, at least on the `FileHandle` instance, but some global cache would be even better. Cache in the `Gdx.files` would make the most sense.

However, I am not sure if I am reading this correctly, but it seems that `list()` on Internal files will always return list of all internal files, regardless of their relation to the path of used `FileHandle`. All other types of FileHandles only return list of files directly under the FileHandle's directory, so this would be extremely confusing.

Reply:
The second part is fixed. It will now no longer return a list of all internal files, just a list of files in the FileHandle it is called on. I'm not sure what you mean by putting it in the cache. I don't know how caching with libgdx works or what its used for.

Reply:
I am also unsure how I would use a BufferedReader on the classpath to read assets, so for now I think I'll keep using Gdx.files.internal because I have already spent too much time on this.